replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/modern/AbstractPersistable.java

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright (c) Connexta
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ditto.replication.api.impl.modern;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import javax.annotation.Nullable;
+import org.codice.ditto.replication.api.modern.Persistable;
+
+/**
+ * This class defines some data that any object to be saved in persistence should have and methods
+ * to manipulate that data.
+ */
+public abstract class AbstractPersistable implements Persistable {
+
+  private static final String ID = "id";
+
+  private String type;
+
+  protected Map<String, Object> properties;
+
+  protected AbstractPersistable(@Nullable String id, String type) {
+    this.type = type;
+    properties = new HashMap<>();
+
+    if (id == null) {
+      setId(UUID.randomUUID().toString());
+    } else {
+      setId(id);
+    }
+  }
+
+  protected AbstractPersistable(String type, Map<String, Object> properties) {
+    this.type = type;
+    setProperties(properties);

```

Comment:
❗️ We shouldn't be calling an abstract method from the constructor. This implies that you would be calling into subclass' code before the subclass has been initialized.

Reply:
I'm not really seeing what might go wrong but I can set the properties in the constructors of the subclasses